### DISCUSSION TITLE
Support for the HTTP Digest authentication on MicroPython (requests and/or aiohttp)

### DISCUSSION THREAD

**Discussion by beyonlo** [micropython/micropython#15138] [Libraries & Drivers]
Hi all!

I need to communicate with some products that have an API that support *only* `HTTP Digest authentication`, I mean, they do not support `HTTP Basic authentication` and it is not possible to disable the authentication. 

I see that MicroPython has two HTTP client libs that can be used for that purpose:
1. `requests` (https://github.com/micropython/micropython-lib/tree/master/python-ecosys/requests) - It is `sync`, and support `HTTP Basic authentication`, but does not support `HTTP Digest authentication`
2. `aiohttp` (https://github.com/micropython/micropython-lib/tree/master/python-ecosys/aiohttp/aiohttp) - It is `async`, and does not support `HTTP Basic authentication` and does not support `HTTP Digest authentication` as well.

I see there is this uwwwauth.py -> https://github.com/pfalcon/pycopy-lib/blob/master/uwwwauth/uwwwauth.py that is a support for `HTTP Digest authentication` for the pycopy (It seems to be like as old fork of MicroPython). Can that help to support `HTTP Digest authentication` on the `requests` and/or `aiohttp` clients?

Is it a simple or difficult task to support `HTTP Digest authentication` on the MicroPython `requests`/`aiohttp`? If I were to choose just one that http client libs to be supported by the `HTTP Digest authentication`, I would like to be the `aiohttp` because it is async (asyncio), so I do not need to use threads to have a non blocking code.

Thank you!

**Comment by jonnor** [comment]
Digest authentication is done using the Authorization header. Both request and aiohttp allow to specify that from the outside. So you do not need to pick a library, nor does this need to happen inside the library. You just need to have code that crafts the correctly formatted header

**Comment by PV-autarky-optimizer** [comment]
Is the question still valid?

I had the same problem. Here is my solution. It connects with 2 pieces of `requests.get(my_url, headers = my_header_handling_the_nonce, auth = None, timeout=20)` and Digest Authentification to MyEnergy API as an example.
[https://myenergi.info/micropython-on-esp32-myenergi-api-t14838.html](https://myenergi.info/micropython-on-esp32-myenergi-api-t14838.html)

However, there are different variants of HTTP Digest authentication. You might need to adapt my solution to your variant. Source code for other variants can be taken from here: [https://github.com/psf/requests/blob/main/src/requests/auth.py](https://github.com/psf/requests/blob/main/src/requests/auth.py)

Good look!
